src/parsers/perf/perfparser.cpp

```diff
@@ -863,24 +863,38 @@ struct PerfParserPrivate
         setParents(&topDownResult.children, nullptr);
     }
 
+    static void buildCallerCalleeResult(const QVector<FrameData>& bottomUpData, QVector<FrameData>& callerCalleeData)
+    {
+        static QSet<QString> callerCalleeRecursionGuard;
```

Comment:
This must not be static, I actually doubt that this algorithm produces the correct results. Please check the current master of heaptrack for how to to this, and test with a recursive test application.

If you need help, let me know. I'll try to find some time later tonight to write a proper unit test for this which you can use then as basis. It's about time I start with unit test coverage, it will make your life much easier.

The problem is the following: The bottom up data contains all call graphs in a "merged" way, but the recursion guard must only be used for individual call graphs. You can get from the "merged"  tree of call stacks to a single stack by doing a depth-first search and then bubbling up the parent chain. I.e. the individual stacks are the parent-chains of every leaf in the bottomUpData tree. Now for every individual stack you'll have to guard against repeated counts of costs in the face of recursion, so you'll clear the set whenever you hit the next leaf before bubbling up the parent chain.

The comment I wrote initially didn't explicitly say that, so sorry if that wasn't clear.